File: The observed PIN.js
var nearby = [[0,8],[1,2,4],[1,2,3,5],[2,3,6],[1,4,5,7],[2,4,5,6,8],[3,5,6,9],[4,7,8],[5,7,8,9,0],[9,8,6]];
function getPINs(observed) {
  var res = [];
  function dfs(path, ind)
  {
    if (ind == observed.length)
    {
      res.push(path.join(''));
      return;
    }
    for (let i = 0; i < nearby[observed[ind]].length; i++)
    {
      path.push(nearby[observed[ind]][i]);
      dfs(path, ind+1);
      path.pop();
    }
  }
  dfs([], 0);
  return res;
}

File: 4-kyu/The observed PIN.py
nearby = [[0,8],[1,2,4],[1,2,3,5],[2,3,6],[1,4,5,7],[2,4,5,6,8],[3,5,6,9],[4,7,8],[5,7,8,9,0],[9,8,6]]

def get_pins(observed):
    res = []
    dfs([], 0, observed, res)
    return res
    
def dfs(path, ind, observed, res):
    if ind == len(observed):
        res += [''.join(str(e) for e in path)]
        return
    for i in xrange(len(nearby[int(observed[ind])])):
        dfs(path + [nearby[int(observed[ind])][i]], ind+1, observed, res)
